code/mob/living/critter/flock/flockdrone.dm

```diff
@@ -370,21 +376,46 @@
 		src.antigrab_counter = 0
 	if(keys & KEY_RUN)
 		if(!src.floorrunning && isfeathertile(src.loc))
-			if(istype(src.loc, /turf/simulated/floor/feather))
+			if (length(src.grabbed_by))
+				for(var/obj/item/grab/g in src.grabbed_by)
+					if (!(g.state == GRAB_PASSIVE || g.state == GRAB_PIN)) // in the rare case you do pin a flockdrone
+						src.can_floorrun = FALSE
+						return ..()
+			else if (!src.can_floorrun)
+				src.can_floorrun = TRUE
+
+			if (istype(src.loc, /turf/simulated/floor/feather))
 				var/turf/simulated/floor/feather/floor = src.loc
+				if (floor.broken)
+					return ..()
 				if(!floor.on)
 					floor.on()
+
 			src.start_floorrunning()
 	else if(keys && src.floorrunning)
 		src.end_floorrunning()
-	. = ..()
+	return ..()
 
 /mob/living/critter/flock/drone/proc/start_floorrunning()
 	if(src.floorrunning)
 		return
 	playsound(src, "sound/misc/flockmind/flockdrone_floorrun.ogg", 50, 1, -3)
 	src.floorrunning = 1
 	src.set_density(0)
+	src.set_pulling(null)
+	if (src.pulled_by)
+		var/mob/M = src.pulled_by
+		M.set_pulling(null)
+	var/obj/item/grab/g = src.find_type_in_hand(/obj/item/grab)
```

Comment:
better pattern might be `for (var/obj/item/grab/G in src.equipped_list())`

Reply:
Only the grip tool for flockdrone can hold items, which is why I went with this, should that still use equipped_list()?

Reply:
It opens it up for expansion in case there was ever something with more than 1 hand slot - not necessary but also it's a small change that might save time/effort down the line